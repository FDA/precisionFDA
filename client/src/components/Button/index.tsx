import React from 'react'
import styled, { css } from 'styled-components'
import { fontWeight, theme } from '../../styles/theme'

export interface IButton {
  active?: "warning" | "danger" | "success" | boolean
  disabled?: boolean
  role?: string
  hide?: boolean
}

export const Button = styled.button<IButton>`
  position: relative;
  display: flex;
  align-items: center;
  font-family: ${theme.fontFamily};
  font-weight: ${fontWeight.regular};
  white-space: nowrap;
  text-align: center;
  background-image: none;
  border: 1px solid transparent;
  box-shadow: 0 2px 0 rgba(0, 0, 0, 0.015);
  cursor: pointer;
  user-select: none;
  touch-action: manipulation;
  line-height: 1.428571429;
  padding: 6px 12px;
  font-size: 14px;
  border-radius: 3px;
  border-color: rgb(218, 239, 251);
  background: white;
  color: ${theme.colors.primaryBlue};

  &:focus {
    outline: 0;
  }

  ${({ role, disabled, active }) => { 
    let textColor = theme.colors.primaryBlue
    let borderColor = 'rgb(218, 239, 251);'
    let borderColorHover = theme.colors.lightBlue
    let backgroundColorHover = theme.colors.subtleBlue

    if (role === "warning") {
      // TODO: the colours are not based on mockups and are not final
      textColor = theme.colors.primaryYellow
      borderColor = theme.colors.darkYellow
      borderColorHover = theme.colors.primaryYellow
      backgroundColorHover = theme.colors.lightYellow
    }
    else if (role === "danger") {
      // TODO: the colours are not based on mockups and are not final
      textColor = theme.colors.primaryRed
      borderColor = theme.colors.darkRed
      borderColorHover = theme.colors.primaryRed
      backgroundColorHover = 'white'
    }
    else if (role === "success") {
      // TODO: the colours are not based on mockups and are not final
      textColor = theme.colors.primaryGreen
      borderColor = theme.colors.darkGreen
      borderColorHover = theme.colors.primaryGreen
      backgroundColorHover = 'white'
    }
    return css`
      color: ${textColor};
      border-color: ${borderColor};
      
      &:hover {
        ${!disabled && css`
          color: ${textColor};
          background: ${backgroundColorHover};
          border-color: ${borderColorHover};
        `}
      }

      &:active {
        color: ${textColor};
        border-color: ${textColor};
        box-shadow: inset 0 3px 5px rgb(0 0 0 / 13%);
      }

      ${!!active && css`
        color: ${textColor};
        border-color: ${textColor};
        box-shadow: inset 0 3px 5px rgb(0 0 0 / 13%);
      `}
    `
  }}

  ${({ disabled }) => disabled && css`
    opacity: 0.5;
    cursor: not-allowed;
  `}

  ${({ hide }) => hide && css`display: none;`}
 `

export const ButtonSolidBlue = styled(Button)`
  background-color: ${theme.colors.primaryBlue};
  border-color: #1b639f;
  color: ${theme.colors.textWhite};

  &:hover {
    ${({ disabled }) => !disabled && css`
      background-color: rgb(24, 85, 137);
      color: ${theme.colors.textWhite};
    `}
    ${({ disabled }) => disabled && css`
      background-color: ${theme.colors.lightBlue};
      cursor: not-allowed;
    `}
  }

  &:active {
    color: ${theme.colors.textWhite};
    background-color: rgb(24, 85, 137);
    box-shadow: inset 0 3px 5px rgb(0 0 0 / 13%);
  }
`

export const ButtonSolidGreen = styled(Button)`
  background-color: ${theme.colors.primaryGreen};
  border-color: ${theme.colors.darkGreen};
  color: ${theme.colors.textWhite};

  &:hover {
    ${({ disabled }) => !disabled && css`
      background-color: ${theme.colors.darkGreen};
      color: ${theme.colors.textWhite};
    `}
    ${({ disabled }) => disabled && css`
      cursor: not-allowed;
    `}
  }

  &:active {
    color: ${theme.colors.textWhite};
    background-color: ${theme.colors.darkGreen};
    box-shadow: inset 0 3px 5px rgb(0 0 0 / 13%);
  }
`

export const ButtonSolidRed = styled(Button)`
  background-color: ${theme.colors.darkRed};
  border-color: ${theme.colors.hoverDarkRed};
  color: ${theme.colors.textWhite};

  &:hover {
    ${({ disabled }) => !disabled && css`
      background-color: ${theme.colors.hoverDarkRed};
      color: ${theme.colors.textWhite};
    `}
    ${({ disabled }) => disabled && css`
      cursor: not-allowed;
    `}
  }

  &:active {
    color: ${theme.colors.textWhite};
    background-color: ${theme.colors.hoverDarkRed};
    box-shadow: inset 0 3px 5px rgb(0 0 0 / 13%);
  }
`

export const ButtonOutlineGrey = styled(Button)`
  border-color: ${theme.colors.borderDefault};
  color: ${theme.colors.textDarkGrey};

  &:hover {
    ${({ disabled }) => !disabled && css`
      background-color: ${theme.colors.backgroundLightGray};
      color: ${theme.colors.textDarkGrey};
      border-color: ${theme.colors.borderDefault};
    `}
    ${({ disabled }) => disabled && css`
      cursor: not-allowed;
    `}
  }

  &:active {
    box-shadow: inset 0 3px 5px rgb(0 0 0 / 13%);
  }
`

export const ToggleButton = styled(Button)<IButton>`
  background-color: ${theme.colors.textWhite};
  border-color: #1b639f;
  color: ${theme.colors.textBlack};

  &:hover {
    ${({ disabled }) => !disabled && css`
      color: ${theme.colors.primaryBlue};
    `}
    ${({ disabled }) => disabled && css`
      color: ${theme.colors.textMediumGrey};
      cursor: not-allowed;
    `}
    ${({ active }) => active && css`
      color: ${theme.colors.textWhite};
      background-color: ${theme.colors.primaryBlue};
    `}
  }

  ${({ active }) => active && css`
    color: ${theme.colors.textWhite};
    background-color: ${theme.colors.primaryBlue};
  `}
`
