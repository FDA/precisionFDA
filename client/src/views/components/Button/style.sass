@import "../../../styles/variables.sass"

.pfda-controller-spaces
  .btn
    position: relative

    &-default
      color: $color-primary
      border-color: $color-primary
