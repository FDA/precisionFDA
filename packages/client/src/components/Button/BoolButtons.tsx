import styled from 'styled-components'
import { Button } from '.'

export const BoolButton = styled(Button)<{'data-selected': BooleanString }>`
  --color: #d9d9d9;
  --darkText: #757575;
  background-color: var(--color);
  border-color: var(--color);
  border: none;
  color: var(--darkText);
  height: 30px;

  &:hover {
    color: var(--darkText);
    border-color: var(--color);
    
    [aria-disabled='true'] {
      color: var(--darkText);
      cursor: not-allowed;
      &[data-selected='true'] {
        color: white;
        background-color: var(--primary-500);
        border-color: var(--primary-500);
      }
    }
  }

  &[data-selected='true'] {
    color: white;
    background-color: var(--primary-500);
    border-color: var(--primary-500);
  }
`

export const BoolButtonGroup = styled.div`
  display: flex;

  &:first-child {
    border-top-left-radius: 2px;
    border-bottom-left-radius: 2px;
    border-right: 0px;
  }
  &:last-child {
    border-top-right-radius: 2px;
    border-bottom-right-radius: 2px;
  }
  button {
    border-radius: 0px;
  }
`
