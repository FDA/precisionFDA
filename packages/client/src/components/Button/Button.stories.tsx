import React from 'react'
import { StoryFn, Meta } from '@storybook/react'
import { Button } from '.'

// More on default export: https://storybook.js.org/docs/react/writing-stories/introduction#default-export
export default {
  title: 'Components/Button',
  component: Button,
  // More on argTypes: https://storybook.js.org/docs/react/api/argtypes
  argTypes: {
    backgroundColor: { control: 'color' },
  },
} as Meta<typeof Button>

// More on component templates: https://storybook.js.org/docs/react/writing-stories/introduction#using-args
const ButtonSolidBlueTemplate: StoryFn<typeof Button> = args => (
  <Button data-variant="primary" {...args}>Submit</Button>
)
const ButtonSolidRedTemplate: StoryFn<typeof Button> = args => (
  <Button data-variant="warning" {...args}>Stop</Button>
)
const ButtonTextTemplate: StoryFn<typeof Button> = args => (
  <Button {...args}>Add a property</Button>
)

export const Blue = ButtonSolidBlueTemplate.bind({})
// More on args: https://storybook.js.org/docs/react/writing-stories/args
Blue.args = {
  disabled: false,
}

export const Red = ButtonSolidRedTemplate.bind({})
Red.args = {
  disabled: false,
}

export const Text = ButtonTextTemplate.bind({})
Text.args = {
  disabled: false,
}
