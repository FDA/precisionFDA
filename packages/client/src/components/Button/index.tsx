import styled, { css } from 'styled-components'

export const Button = styled.button.attrs(({ disabled }) => ({ 'aria-disabled': disabled ? true : undefined }))<{
  'data-variant'?: 'primary' | 'success' | 'warning' | 'link'
  disabled?: boolean
  active?: BooleanString
} >`
  position: relative;
  display: flex;
  width: fit-content;
  align-items: center;
  white-space: nowrap;
  text-align: center;
  background-image: none;
  border: 1px solid transparent;
  box-shadow: 0 2px 0 rgba(0, 0, 0, 0.015);
  cursor: pointer;
  user-select: none;
  touch-action: manipulation;
  line-height: 1.4285;
  padding: 6px 12px;
  font-size: 14px;
  border-radius: 3px;
  border-color: var(--tertiary-300);
  background: var(--tertiary-100);
  transition: all 0.1s;

  &:active {
    box-shadow: inset 0 3px 5px rgba(0, 0, 0, 0.2);
  }

  &:hover {
    filter: brightness(97%);
  }

  &:focus {
    outline: 0;
  }

  &[data-variant='primary'] {
    border-color: var(--primary-600);
    background: var(--primary-500);
    color: hsl(0, 0%, 97%);
    &:hover {
      filter: brightness(94%);
    }
    &:active {
      box-shadow: inset 0 3px 5px rgba(0, 0, 0, 0.26);
      filter: brightness(93%);
      border-color: var(--primary-500);
    }
  }

  &[data-variant='warning'] {
    border-color: var(--warning-600);
    background: var(--warning-500);
    color: hsl(0, 0%, 97%);
    &:hover {
      filter: brightness(94%);
    }
    &:active {
      box-shadow: inset 0 3px 5px rgba(0, 0, 0, 0.26);
      filter: brightness(93%);
      border-color: var(--warning-500);
    }
  }

  &[data-variant='success'] {
    border-color: var(--success-700);
    background: var(--success-600);
    color: hsl(0, 0%, 97%);
    &:hover {
      filter: brightness(94%);
    }
    &:active {
      box-shadow: inset 0 3px 5px rgba(0, 0, 0, 0.26);
      filter: brightness(93%);
      border-color: var(--success-600);
    }
  }

  &[data-variant='link'] {
    background: none;
    border: none;
    box-shadow: none;
    color: var(--c-link);
    &:hover {
      color: var(--c-link-hover);
    }
    &:active {
      color: var(--c-link-hover);
    }
  }

  &[aria-disabled] {
    cursor: not-allowed;
    color: var(--c-text-400);
    border: 1px solid transparent;
    background-color: var(--tertiary-200);
    &:hover {
      filter: initial;
      cursor: not-allowed;
    }
    &:active {
      box-shadow: initial;
      border: 1px solid transparent;
    }
  }
`

export const TransparentButton = styled(Button)`
  display: flex;
  box-shadow: none;
  background: none;
  border: none;
  cursor: pointer;
  padding: 0;
  &:active {
    position: relative;
    top: 1px;
    outline: none;
    box-shadow: none;
    filter: none;
  }
  &:focus {
    outline: none;
  }
`

export const buttonDarkModeHack = css``
