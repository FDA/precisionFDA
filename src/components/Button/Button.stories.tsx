import React from 'react'
import { StoryFn, Meta } from '@storybook/react'

// import { Button } from './Button';
import {
  Button,
  ButtonSolidBlue,
  ButtonSolidGreen,
  ButtonSolidRed,
} from '.'

// More on default export: https://storybook.js.org/docs/react/writing-stories/introduction#default-export
export default {
  title: 'Components/Button',
  component: Button,
  // More on argTypes: https://storybook.js.org/docs/react/api/argtypes
  argTypes: {
    backgroundColor: { control: 'color' },
  },
} as Meta<typeof Button>

// More on component templates: https://storybook.js.org/docs/react/writing-stories/introduction#using-args
const ButtonSolidBlueTemplate: StoryFn<typeof Button> = args => (
  <ButtonSolidBlue {...args}>Submit</ButtonSolidBlue>
)
const ButtonSolidRedTemplate: StoryFn<typeof Button> = args => (
  <ButtonSolidRed {...args}>Stop</ButtonSolidRed>
)
const ButtonSolidGreenTemplate: StoryFn<typeof Button> = args => (
  <ButtonSolidGreen {...args}>All Good</ButtonSolidGreen>
)

export const Blue = ButtonSolidBlueTemplate.bind({})
// More on args: https://storybook.js.org/docs/react/writing-stories/args
Blue.args = {
  disabled: false,
}

export const Red = ButtonSolidRedTemplate.bind({})
Red.args = {
  disabled: false,
}

export const Green = ButtonSolidGreenTemplate.bind({})
Green.args = {
  disabled: false,
}
