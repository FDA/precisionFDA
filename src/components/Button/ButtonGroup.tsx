import styled from 'styled-components'
import { ToggleButton } from '.'

// Use ToggleButton to implement a multi-selection button group

export const ButtonGroup = styled.div`
  display: flex;

  ${ToggleButton}{
    border-radius: 0;
    border-right: 0px;
  }

  ${ToggleButton}:first-child {
    border-top-left-radius: 2px;
    border-bottom-left-radius: 2px;
  }
  ${ToggleButton}:last-child {
    border-top-right-radius: 2px;
    border-bottom-right-radius: 2px;
    border-right: 1px;
  }
`
