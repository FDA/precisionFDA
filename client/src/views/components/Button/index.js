import React from 'react'
import PropTypes from 'prop-types'
import classNames from 'classnames/bind'

import './style.sass'


const TYPES = ['danger', 'success', 'warning', 'primary', 'info', 'default']
const SIZES = ['sm', 'xs', 'lg']

const Button = ({ children, type, size, ...rest }) => {
  const typeClass = TYPES.includes(type) ? `btn-${type}` : 'btn-default'
  const sizeClass = SIZES.includes(size) ? `btn-${size}` : null

  const classes = classNames(['btn', typeClass, sizeClass])
  return <button {...rest} className={classes}>{children}</button>
}

export default Button

Button.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.element,
    PropTypes.arrayOf(PropTypes.element),
  ]),
  type: PropTypes.oneOf(TYPES),
  size: PropTypes.oneOf(SIZES),
}
