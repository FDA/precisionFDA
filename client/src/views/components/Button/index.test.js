import React from 'react'
import { shallow } from 'enzyme'

import Button from './index'


describe('Button', () => {
  it('matches snapshot', () => {
    const wrapper = shallow(<Button>Click me</Button>)

    expect(wrapper).toMatchSnapshot()
  })

  describe('when predefined size given', () => {
    it('has predefined size class', () => {
      const wrapper = shallow(<Button size="xs">Click me</Button>)

      expect(wrapper).toHaveClassName('btn-xs')
    })
  })

  describe('when non-predefined size given', () => {
    it('has no size class', () => {
      const wrapper = shallow(<Button size="some-size">Click me</Button>)

      expect(wrapper).not.toHaveClassName('btn-xs')
      expect(wrapper).not.toHaveClassName('btn-sm')
      expect(wrapper).not.toHaveClassName('btn-lg')
    })
  })

  describe('when predefined type given', () => {
    it('has predefined type class', () => {
      const wrapper = shallow(<Button type="info">Click me</Button>)

      expect(wrapper).toHaveClassName('btn-info')
    })
  })

  describe('when non-predefined type given', () => {
    it('doesn\'t default type class', () => {
      const wrapper = shallow(<Button type="some-type">Click me</Button>)

      expect(wrapper).toHaveClassName('btn-default')
    })
  })
})
