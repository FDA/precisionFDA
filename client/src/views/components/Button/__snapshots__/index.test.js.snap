// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button matches snapshot 1`] = `
<button
  className="btn btn-default"
>
  Click me
</button>
`;
